package/metrics_daemon/src/metrics_daemon.c

```diff
@@ -0,0 +1,349 @@
+            }
+        }
+        pdest = pdest2; ///< if there is still sub-folders, update the pdest to pdest2
+    }
+    return json_current;
+}
+
+
+/**
+ * @brief handle_walk_path - function get called by ftw for each of the sub-folder or file under the path
+ *
+ * @param const char *fpath: file path in ASCII
+ * @param const struct stat *sb: file status struct
+ * @param int tflag: flag indicates file/folder type
+ * @return int : integer that indicates if the walk through failed.
+ */
+static int handle_walk_path(const char *fpath, const struct stat *sb, int tflag)
+{
+    if(first_folder == TRUE)     ///< skip the root since it is handled by main()
+    {
+        first_folder = FALSE;
+        return 0;
+    }
+    char * bname = extract_filename(fpath);
+    json_object * jobj_Cur = loop_through_folder_name(fpath,root_name, root_length, jobjRoot);  ///< get the json object needs to be updated
+    switch(tflag)
+    {
+        case FTW_D:
+        {
+            bool file_entry_exist = FALSE;
+            json_object_object_foreach(jobj_Cur, key, val) {
+                if(strcmp(bname,key)==0)
+                {
+                    file_entry_exist = TRUE;
+                    jobjCur = val;
+                    break;
+                }
+            }
+            if(!file_entry_exist)                 ///< if it is a folder and it doesn't exist in json tree, add a new folder, do nothing otherwise
+            {
+                json_object *jnewobj = json_object_new_object();     ///< Create a new object here to set the name of folder
+                json_object_object_add(jobj_Cur,bname, jnewobj);
+            }
+            break;
+        }
+
+        case FTW_F:
+        {
+            FILE * fp = fopen(fpath, "r");           ///< read file to get the data
+            char buf[64];
+            long file_len_long = sb->st_size-1;
+            int file_len = (int)(sb->st_size-1);
+            if(file_len<64 && file_len_long >=0)
+            {
+                fread(&buf[0],(unsigned int)file_len,1,fp);
+            } else{
+                file_len = 0;                        ///< if file is empty, set the str length to zero
+            }
+            fclose(fp);
+            bool file_entry_exist = FALSE;
+            json_object_object_foreach(jobj_Cur, key, val) {
+                if(strcmp(bname,key)==0)
+                {
+                    file_entry_exist = TRUE;
+                    json_object_set_string_len(val,&buf[0],file_len);     ///< if file exist, update
+                    break;
+                }
+            }
+            if(!file_entry_exist)     ///< add a new node if it doesn't exist
+            {
+                json_object_object_add(jobj_Cur,bname, json_object_new_string_len(&buf[0],file_len));
+            }
+            break;
+        }
+        default:
+        {
+            perror("FTW: unexpected file type\n");   ///< For now, only support file and forlder
```

Comment:
Calling `perror` will print errno with the message supplied, is this intentional?  If this is an expected scenario, just silently ignore, otherwise use `piksi_log` to log an error.  Also this path leaks `bname`.